model/perf.go

```diff
@@ -637,3 +657,142 @@ func (r *PerformanceResults) FindOutdatedRollups(ctx context.Context, name strin
+					"project":     "$info.project",
+					"test":        "$info.test_name",
+					"measurement": "$rollups.stats.name",
+				},
+			},
+		},
+	}
+	return db.Collection(perfResultCollection).Aggregate(ctx, pipeline)
+}
+
+type TimeSeriesId struct {
+	Project     string `bson:"project"`
+	Variant     string `bson:"variant"`
+	Task        string `bson:"task"`
+	Test        string `bson:"test"`
+	Measurement string `bson:"measurement"`
+}
+
+type TimeSeriesEntry struct {
+	PerfResultID string  `bson:"_id"`
+	Value        float64 `bson:"value"`
+	Order        int     `bson:"order"`
+}
+
+type TimeSeries struct {
+	TimeSeriesId TimeSeriesId      `bson:"_id"`
+	Data         []TimeSeriesEntry `bson:"times_series"`
+}
+
+func makeTimeSeriesPipeline(timeSeriesId TimeSeriesId) []bson.M {
+	return []bson.M{
+		{
+			"$match": bson.M{
+				"info.project":   timeSeriesId.Project,
+				"info.variant":   timeSeriesId.Variant,
+				"info.task_name": timeSeriesId.Task,
+				"info.test_name": timeSeriesId.Test,
+			},
+		},
+		{
+			"$match": bson.M{
+				"info.order": bson.M{
+					"$exists": true,
+				},
+				"info.mainline": true,
+			},
+		},
+		{
+			"$unwind": "$rollups.stats",
+		},
+		{
+			"$match": bson.M{
+				"rollups.stats.name": timeSeriesId.Measurement,
+			},
+		},
+		{
+			"$project": bson.M{
+				"value": "$rollups.stats.val",
+				"order": "$info.order",
+			},
+		},
+	}
+}
+
+func GetTimeSeries(ctx context.Context, env cedar.Environment, timeSeriesId TimeSeriesId) (*TimeSeries, error) {
+	timeSeriesAggregator := makeTimeSeriesPipeline(timeSeriesId)
+	cur, err := env.GetDB().Collection(perfResultCollection).Aggregate(ctx, timeSeriesAggregator)
+	if err != nil {
+		return nil, errors.Wrap(err, "Cannot aggregate time series")
+	}
+	defer cur.Close(ctx)
+	var res []TimeSeriesEntry
+	err = cur.All(ctx, &res)
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not decode time series")
+	}
+	return &TimeSeries{
+		TimeSeriesId: timeSeriesId,
+		Data:         res,
+	}, nil
+}
+
+func ClearChangePoints(ctx context.Context, env cedar.Environment, timeSeriesId TimeSeriesId) error {
+	seriesFilter := bson.M{
+		"info.project":              timeSeriesId.Project,
+		"info.variant":              timeSeriesId.Variant,
+		"info.task_name":            timeSeriesId.Task,
+		"info.test_name":            timeSeriesId.Test,
+		"rollups.stats.name":        timeSeriesId.Measurement,
+		"change_points.measurement": timeSeriesId.Measurement,
+	}
+	pullUpdate := bson.M{
+		"$pull": bson.M{
+			"change_points": bson.M{
+				"measurement": timeSeriesId.Measurement,
+			},
+		},
+	}
+	_, err := env.GetDB().Collection(perfResultCollection).UpdateMany(ctx, seriesFilter, pullUpdate)
+	return errors.Wrap(err, "Unable to clear change points")
+}
+
+func CreateChangePoint(ctx context.Context, env cedar.Environment, resultToUpdate string, measurement string, algorithm AlgorithmInfo) error {
+	filter := bson.M{"_id": resultToUpdate}
+	update := bson.M{
+		"$push": bson.M{
```

Comment:
does this make idempotency, upon re-run, more complicated?

Reply:
I think any change points that match from previous jobs get cleared out immediately prior to this. So re-running would be clear-->add one-->clear-->add one-->clear-->add one. Seems idempotent to me.

Reply:
per our in person conversation, we think it makes sense to combine these operations into a single operation,.